query-algebrizer/src/clauses/mod.rs

```diff
@@ -825,7 +826,139 @@ impl ConjoiningClauses {
     /// where `?y` must simultaneously be a ref-typed attribute and a boolean-typed attribute. This
     /// function deduces that and calls `self.mark_known_empty`. #293.
     pub fn expand_type_tags(&mut self) {
-        // TODO.
+        for (v, tag) in self.extracted_types.iter() {
+            let cols = self.column_bindings.get(v).unwrap().clone();
+            for col in cols {
+                if col.0 != tag.0 { 
+                    let value = QueryValue::Column(QualifiedAlias(col.0.clone(), Column::Fixed(DatomsColumn::ValueTypeTag)));
+                    let alias = tag.clone();
+                    self.wheres.0.push(ColumnConstraintOrAlternation::Constraint(ColumnConstraint::Equals(alias, value)));
+                }
+            }
+        }
+
+        let mut computed_types: BTreeMap<Variable, TypedValue> = BTreeMap::default();
+
+        for computed_table in self.computed_tables.iter() {
+            match computed_table {
+                &ComputedTable::NamedValues { ref names, ref values } => {
+                    for (i, item) in names.iter().enumerate() {
+                        computed_types.insert(item.clone(), values[i].clone());
+                    }
+                },
+                _ => (),
+            }
+        }
+
+        let mut empty_reason: Option<EmptyBecause> = None;
+        let mut wheres_to_add = vec![];
+        for (v, value_type_set) in self.known_types.iter() {
+            if self.column_bindings.contains_key(v) {
+                let cols = self.column_bindings.get(v).unwrap().clone();
+                for col in cols {
+                    match col.1 {
+                        Column::Fixed(DatomsColumn::Value) => { 
+                            // attempt to figure out exactly which type is used
+                            // TODO: But only if the attribute is a variable or not present and therefore of unknown type
+                            let mut value_type: Option<ValueType> = None;
+                            for constraint in self.wheres.0.iter() {
```

Comment:
Question:

In this example:

```
[:find ?x 
 :where [?x ?y true]
        [?z ?y ?x]]
```

we get the cc:

```
ConjoiningClauses { 
    empty_because: None, 
    from: [
        SourceAlias(Datoms, datoms00), 
        SourceAlias(Datoms, datoms01)
    ], 
    computed_tables: [], 
    wheres: [
        Constraint(datoms00.v = value(Boolean(true))), 
        Constraint("datoms00".value_type_tag = Boolean), 
        Constraint(datoms00.e = datoms01.v), 
        Constraint(datoms00.a = datoms01.a)
    ], 
    column_bindings: {
        var(?x): [datoms00.e, datoms01.v], 
        var(?y): [datoms00.a, datoms01.a], 
        var(?z): [datoms01.e]
    }, 
    input_variables: {}, 
    value_bindings: {}, 
    known_types: {
        var(?x): ValueTypeSet({Ref}), 
        var(?y): ValueTypeSet({Ref}), 
        var(?z): ValueTypeSet({Ref})
    }, 
    extracted_types: {} 
}
```

`var(?y)` has a known type of `Ref` but has a `bool` assigned to it. This should cause a `TypeMismatch` `EmptyBecause` to be attached. 

Why is it that the known types for `var(?y)` does not contain `Boolean`? I would have thought we could determine that, but we only determine the `Ref` type. 

Is this a bug? 

If the known types for `var(?y)` included `Boolean` then that would make everything a lot easier in this case...